lib/Debugger.ProtocolHandler/ProtocolHandler.cpp

```diff
@@ -125,16 +125,16 @@ namespace JsDebug
     void ProtocolHandler::sendProtocolNotification(std::unique_ptr<Serializable> message)
     {
         protocol::String str = message->serialize();
-        std::string asciiStr = str.toAscii();
-        const char* response = asciiStr.c_str();
 
 #ifdef _DEBUG
+        static_assert(sizeof(wchar_t) == sizeof(uint16_t));
         OutputDebugStringA("{\"type\":\"response\",\"payload\":");
-        OutputDebugStringA(response);
+        OutputDebugStringW(reinterpret_cast<const wchar_t*>(str.characters16()));
         OutputDebugStringA("},\r\n");
 #endif
 
-        SendResponse(response);
+        std::string utf8Str = str.toUtf8();
+        SendResponse(utf8Str.c_str());
```

Comment:
If utf8 conversion fails here, returning an empty string might cause unpredictable problems. Seems like removing bad characters would be a better approach so protocol messages would remain mostly intact (with the assumption that a bad char would be infrequent and likely coming from user data rather than the protocol itself).

Reply:
From https://docs.microsoft.com/en-us/windows/desktop/api/stringapiset/nf-stringapiset-widechartomultibyte:

WC_ERR_INVALID_CHARS | Windows Vista and later: Fail (by returning 0 and setting the last-error code to  ERROR_NO_UNICODE_TRANSLATION) if an invalid input character is encountered. You can retrieve the last-error code with a call to GetLastError. If this flag is not set, the function replaces illegal sequences with U+FFFD (encoded as appropriate for the specified codepage) and succeeds by returning the length of the converted string. Note that this flag only applies when CodePage is specified as CP_UTF8 or 54936. It cannot be used with other code page values.
-- | --

It looks like the default behavior for UTF-8 is to just insert a replacement character for any illegal sequences.  I believe that's functionally equivalent to just dropping the character.

Reply:
It's actually better to put the replacement character rather than dropping them. The string length then remains the same and you can see that there is a character problem.